WDtCCtR Silver P2: Upgrade explanation search to N^2 to match code, add tips on actual N^2 impl (#6245)
cpinitiative/usaco-guide: solutions/silver/usaco-716.mdx

---
id: usaco-716
source: USACO Silver 2017 February
title: Why Did the Cow Cross the Road III
author: Óscar Garries, Kevin Sheng, David Guo, John Zhou
---

[Official Analysis (Java)](http://www.usaco.org/current/data/sol_countcross_silver_feb17.html)

## Explanation

Let's view the grid as a graph, where each grid is a vertex, and each border between two grids that does not contain a road be an edge.

> For two cows to need to cross a road to reach each other, they must belong to two different connected components in the graph.

Reviewer: ```suggestion
Let's view the grid as a graph, where each cell is a vertex, and each border between two grids that does not contain a road is an edge.
For two cows to need to cross a road to reach each other, they must belong to two different connected components in the graph.
```

We first perform BFS or DFS from an arbitrary point in each connected component to count the number of cows in it.
Now, for each pair of connected components $A$ and $B$, any cow from $A$ must cross a road to reach any cow from $B$;

> therefore, the answer is the sum of the pairwise products of the number of cows in each connected component, taking special care to not count double for each pair.

Reviewer: ```suggestion
We perform floodfill to find each connected component and the number of cows within it. For a pair of connected components $A$ and $B$, each possible pair of cows must cross a road to reach the other. Thus, the answer is the sum of pairwise products between the number of cows in each component.
```

or smth like this, the current one is a little choppy/reads unnaturally

Author: > We perform floodfill to find each connected component and the number of cows within it. For a pair of connected components $A$ and $B$, each possible pair of cows must cross a road to reach the other. Thus, the answer is the sum of pairwise products between the number of cows in each component.

```suggestion
We perform floodfill to find each connected component and the number of cows within it.
For a pair of connected components, each possible pair of cows between these connected components must cross a road to reach the other.
Thus, the answer is the sum of pairwise products between the number of cows in each component.
```

> <Spoiler title="Implementation Details for the Pairwise Product Sum">

Reviewer: ok so rn this spoiler doesn't really make sense (like the title/when reading the article)

maybe preface this by saying "Naively implementing the solution above takes $\mathcal{O}(N^4)$ time, as we are computing pairwise products with up to $\mathcal{O}(N^2)$ components. However, we can optimize this to $\mathcal{O}(N^2)$."

and then name the spoiler title like, "Optimizing Pairwise Product Sums"

Author: ```suggestion
Naively implementing this solution takes $\mathcal{O}(N^4)$ time, as we are computing pairwise products with up to $\mathcal{O}(N^2)$ components; this can be optimized to $\mathcal{O}(N^2)$:

<Spoiler title="Optimizing Pairwise Product Sums">
```

The BFS algorithm takes $\mathcal{O}(N^2)$ for a floodfill; however, the time complexity will still be $\mathcal O(N^4)$ because of a bottleneck:
Since there will be at most $N^2$ connected components, the time complexity of computing the pairwise product of the counts can be $\mathcal{O}(N^4)$.
This is fast enough to pass; however, there is a way to get rid of this bottleneck:

Let $k$ be the total number of connected components, and $\mathrm{C}[i]$ ($1\leq i\leq k$) be the number of cows in the $i$-th connected component.
Naïvely, we would compute $\sum_{i=1}^{k}\sum_{j=1}^{i-1}\mathrm{C}[i]\mathrm{C}[j]$ for our answer;
however, we can precompute $\mathrm{C_{sum}}[i] = \sum_{j=1}^{i}\mathrm{C}[j]$ in $\mathcal{O}(k)$ using prefix sums, and then compute

> $\sum_{i=1}^{k}\sum_{j=1}^{i-1}\mathrm{C}[i]\mathrm{C}[j]=\sum_{i=1}^{k}\mathrm{C}[i]\sum_{j=1}^{i-1}\mathrm{C}[j]=\sum_{i=1}^{k}\mathrm{C}[i]\mathrm{C_{sum}}[i-1]$ in $\mathcal{O}(k)$ time as well.  To simplify the implementation, we could also use a single loop, and keep track of $\mathrm{C_{sum}}[i-1]$ by accumulating $\mathrm{C}[i]$ to it cumulatively after we use the sum for our main computation.

Reviewer: <img width="859" height="364" alt="Image" src="https://github.com/user-attachments/assets/f15d34a0-ce65-4829-bdf3-b8dd3c9be2e3" />

IMO this is not very readable as it is

how i would explain it is with using more pseudocode

```cpp
vector<int> sizes(num_comps);
int sum = 0;
for (int i = 0; i < num_comps; i++) {
    for (int j = 0; j < i; j++) {
        sum += sizes[i] * sizes[j];
    }
}
```

then just say "oh by the distributive property we can just have a running sum gg" and put the optimized vers

Author: @eysbutno What language should pseudocode be in?x

For two cows to need to cross a road to reach each other, they must belong to two different connected components in the graph.

We perform floodfill to find each connected component and the number of cows within it.
For a pair of connected components, each possible pair of cows between these connected components must cross a road to reach the other.
Thus, the answer is the sum of pairwise products between the number of cows in each component.

Naively implementing this solution takes $\mathcal{O}(N^4)$ time, as we are computing pairwise products with up to $\mathcal{O}(N^2)$ components; this can be optimized to $\mathcal{O}(N^2)$:

<Spoiler title="Optimizing Pairwise Product Sums">

The BFS algorithm takes $\mathcal{O}(N^2)$ for a floodfill; however, the time complexity will still be $\mathcal O(N^4)$ because of a bottleneck:
Since there will be at most $N^2$ connected components, the time complexity of computing the pairwise product of the counts can be $\mathcal{O}(N^4)$.
This is fast enough to pass; however, there is a way to get rid of this bottleneck:

> 

Author: ```suggestion
```
Gettting rid of this as the part before the spoiler already explains this well.

therefore, the answer is the sum of the pairwise products of the number of cows in each connected component, taking special care to not count double for each pair.

<Spoiler title="Optimizing the time complexity">

The BFS algorithm takes $\mathcal{O}(N^2)$ for a floodfill; however, the time complexity will still be $\mathcal O(N^4)$ because of a bottleneck:
Since there will be at most $N^2$ connected components, the time complexity of computing the pairwise product of the counts can be $\mathcal{O}(N^4)$.
This is fast enough to pass; however, there is a way to get rid of this bottleneck:

Let $k$ be the total number of connected components, and $\mathrm{C}[i]$ ($1\leq i\leq k$) be the number of cows in the $i$-th connected component.
Naïvely, we would compute $\sum_{i=1}^{k}\sum_{j=i+1}^{k}\mathrm{C}[i]\mathrm{C}[j]$ for our answer;
however, we can precompute $\mathrm{C_{sum}}[i] = \sum_{j=i}^{k}\mathrm{C}[j]$ in $\mathcal{O}(k)$ using postfix sums, and then compute
$\sum_{i=1}^{k}\sum_{j=i+1}^{k}\mathrm{C}[i]\mathrm{C}[j]=\sum_{i=1}^{k}\mathrm{C}[i]\sum_{j=i+1}^{k}\mathrm{C}[j]=\sum_{i=1}^{k}\mathrm{C}[i]\mathrm{C_{sum}}[i+1]$ in $\mathcal{O}(k)$ time as well.

> This is technically not needed to pass the problem as-is, so for simplicity, it is not implemented in the code below.

Reviewer 2: isn't it simple to implement though?

Author: @bqi343 orz orz orz

Yeah... I was kind of in the mentality of giving the reader an exercise to implement... I'll try to add an implementation if it's needed.

		}
	}

	// +1 here so that an element after the last one can be 0, which simplifies the
	// implementation as we can start adding onto that 0.
	vector<int> cow_components_postfix(cow_components.size() + 1);
	for (int i = cow_components.size() - 1; i >= 0; i--) {
		cow_components_postfix[i] = cow_components_postfix[i + 1] + cow_components[i];
	}

	int distant_pairs = 0;
	for (int i = 0; i < cow_components.size(); i++) {

> 		distant_pairs += cow_components[i] * cow_components_postfix[i + 1];

Reviewer 2: isn't it simpler to just maintain one integer with  the running sum that you update as i changes, rather than requiring an entire vector

Author: @bqi343 Well it's the sum of all elements after i, so introducing another integer, summing all the numbers, and then subtracting cow_components[i] as we go is more *conceptual* complexity, even though it's more implementationally complex to do it the current way.

What do you think?

Reviewer 2: you can do prefix instead of postfix to avoid the subtraction

Author: Duh.  Thanks!
